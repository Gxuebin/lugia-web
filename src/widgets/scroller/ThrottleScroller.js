/**
 * 节流滚动HOC
 * create by ligx
 *
 * @flow
 */
import * as React from 'react';
import styled from 'styled-components';
import Scroller from './index';
import Widget from '../consts/index';
import { FontSizeNumber } from '../css';
import { BarDefaultSize, DefaultHeight, DefaultWidth } from '../css/scroller';
import { getCanSeeCount } from './support';
import { px2emcss } from '../css/units';

const em = px2emcss(FontSizeNumber);

const height = props => {
  const { theme, totalSize, autoHeight = false } = props;
  const { height: themeHeight } = theme;
  if (!autoHeight) {
    return themeHeight ? `height:${em(themeHeight)};` : `height:${em(DefaultHeight)};`;
  }

  if (!themeHeight || themeHeight > totalSize) {
    return `height: ${em(totalSize)}`;
  }
  return `height: ${em(themeHeight)}`;
};

const getActiveWidth = props => {
  const { theme } = props;
  const { width } = theme;
  return width;
};

const width = props => {
  const width = getActiveWidth(props);
  return width ? `width:${em(width)};` : `width:${em(DefaultWidth)};`;
};
const getContentWidthValue = props => {
  const width = getActiveWidth(props);
  return width ? width - BarDefaultSize : DefaultWidth - BarDefaultSize;
};

const getContentWidth = props => {
  const width = getContentWidthValue(props);
  return em(width);
};

const scrollerLeft = props => {
  return `left: ${getContentWidth(props)};`;
};

const Col = styled.div`
  ${width};
  font-size: ${FontSizeNumber}rem;
  position: absolute;
  display: inline-block;
`;

const ScrollerContainer = styled.div`
  overflow: hidden;
  font-size: ${FontSizeNumber}rem;
  ${height};
  ${width};
  position: relative;
`;

const getOpacity = (props: Object) => {
  const { isDrag } = props;
  if (isDrag) {
    return 'opacity: 1;';
  }
  return '';
};
const ScrollerCol = Col.extend`
  ${scrollerLeft};
  width: ${em(BarDefaultSize)};
  opacity: 0;
  ${ScrollerContainer}:hover & {
    opacity: 1;
  }
  ${getOpacity};
  transition: opacity 0.3s;
`;

type ThrottleScrollerState = {
  start: number,
};
export default (Target: React.ComponentType<any>, MenuItemHeight: number) => {
  return class ThrottleScroller extends React.Component<any, ThrottleScrollerState> {
    static displayName = Widget.ThrottleScroller;
    static defaultProps = {
      mutliple: false,
      getTheme: () => {
        return {};
      },
    };
    scroller: ?Object;
    scrollerTarget: ?Object;

    itemHeight: number;

    isDrag: boolean;

    constructor(props: any) {
      super(props);
      this.state = {
        start: 0,
      };
      this.itemHeight = this.getActiveItemHeight(props);
    }

    getActiveItemHeight = (props: Object) => {
      const { menuItemHeight } = props;
      const menuItemHeightIsInProps = 'menuItemHeight' in props;
      return menuItemHeightIsInProps ? menuItemHeight : MenuItemHeight;
    };

    render() {
      const { props } = this;

      const start = this.getStart(props, this.state);
      const { getTheme } = props;
      const theme = getTheme();
      const { level, autoHeight = false } = props;
      const totalSize = this.fetchTotalSize();

      const pack = (element: Object | Array<Object>) => {
        return (
          <ScrollerContainer
            level={level}
            totalSize={totalSize}
            theme={theme}
            autoHeight={autoHeight}
            onWheel={this.onWheel}
          >
            {element}
          </ScrollerContainer>
        );
      };

      if (!this.isNeedScroller()) {
        const { length } = this.getTarget();
        //TODO: 待测试
        return pack(
          <Target
            {...props}
            start={0}
            end={length}
            canSeeCount={length}
            ref={cmp => (this.scrollerTarget = cmp)}
          />
        );
      }

      const { type, step } = props;
      const viewSize = this.fetchViewSize();

      const end = this.fetchEnd(start);
      const canSeeCount = this.canSeeCount();
      const menuItemHeight = this.itemHeight;
      return pack([
        <Col theme={theme} level={level}>
          <Target
            {...props}
            canSeeCount={canSeeCount}
            start={start}
            end={end}
            ref={cmp => (this.scrollerTarget = cmp)}
          />
        </Col>,
        <ScrollerCol level={level} theme={theme} isDrag={this.isDrag}>
          <Scroller
            ref={cmp => (this.scroller = cmp)}
            type={type}
            onDrag={this.onDrag}
            value={menuItemHeight * start}
            viewSize={viewSize}
            totalSize={totalSize}
            onChange={this.onScroller}
            step={step}
          />
        </ScrollerCol>,
      ]);
    }

    onDrag = (drag: boolean) => {
      this.isDrag = drag;
    };

    getStart(props: Object, state: Object): number {
      const { length } = this.getTarget();

      const limitStart = (start: number) => {
        const maxStart = Math.max(0, length - this.canSeeCount() + 1);
        return Math.min(start, maxStart);
      };

      if ('start' in props) {
        const { start = 0 } = props;
        return limitStart(start);
      }

      const { start = 0 } = state;
      return limitStart(start);
    }

    isNeedScroller() {
      const { length } = this.getTarget();
      return this.canSeeCount() < length;
    }

    canSeeCount(): number {
      return getCanSeeCount(this.fetchViewSize(), this.itemHeight);
    }

    fetchViewSize = () => {
      const { autoHeight = false } = this.props;
      const { height: themeHeight } = this.props.getTheme();
      if (!autoHeight) {
        return themeHeight || themeHeight === 0 ? themeHeight : DefaultHeight;
      }
      const { data } = this.props;
      const allItemHeight = this.itemHeight * data.length;
      if (!themeHeight || themeHeight > allItemHeight) {
        return allItemHeight;
      }
      return themeHeight;
    };

    fetchTotalSize(): number {
      const { length } = this.getTarget();
      return length * this.itemHeight;
    }

    getTarget(): Array<any> {
      const { data, children } = this.props;
      if (!data && !children) {
        return [];
      }
      return data ? data : children;
    }

    fetchEnd(start: number): number {
      const { length: maxLen } = this.getTarget();
      if (maxLen === 0) {
        return 0;
      }
      const seeCount = this.canSeeCount();

      return Math.min(seeCount + Math.max(start, 0), maxLen);
    }

    onWheel = (e: Object) => {
      if (this.isNeedScroller()) {
        e.stopPropagation();
        e.preventDefault();
      }
      this.scroller && this.scroller.onWheel && this.scroller.onWheel.call(this.scroller, e);
    };

    onScroller = (value: number) => {
      const { onScroller } = this.props;
      const start = value / this.itemHeight;
      onScroller ? onScroller(start, this.fetchEnd(start)) : this.setState({ start });
    };
  };
};
