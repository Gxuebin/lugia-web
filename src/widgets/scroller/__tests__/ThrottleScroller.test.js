//@flow

import * as React from 'react';
import 'jest-styled-components';
import Enzyme, { mount, render, } from 'enzyme';
import { createTestComponent, } from 'sv-test-utils';
import Adapter from 'enzyme-adapter-react-16';
import chai from 'chai';
import ThrottleScroller from '../ThrottleScroller';
import { mockObject, } from 'vx-mock';
import renderer from 'react-test-renderer';

import SVScroller from '../';


const { expect: exp, } = chai;

Enzyme.configure({ adapter: new Adapter(), });
describe('ThrottleScroller', function () {

  it('getTarget: target is data', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10,];
    const Target = createTestComponent(ThrottleScroller(Menus, 20), target => {
      exp(target.getTarget()).to.be.equal(data);
    });
    mount(<Target data={data}/>);
  });
  it('getTarget: target is children', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const Item = (props: Object) => {
      return <div>{props.children}</div>;
    };
    const myItems = [<Item key="a">hello</Item>, <Item key="b">world</Item>,];
    const Target = createTestComponent(ThrottleScroller(Menus, 20), target => {
      exp(target.getTarget()).to.be.equal(myItems);
    });
    mount(<Target>{myItems}</Target>);
  });
  it('getTarget: target is null', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const Target = createTestComponent(ThrottleScroller(Menus, 20), target => {
      exp(target.getTarget()).to.be.eql([]);
    });
    mount(<Target></Target>);
  });

  it('isNeedScrolelr is false  data的数量为0', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const Target = createTestComponent(ThrottleScroller(Menus, 20), target => {
      const mock = mockObject.create(target);
      const getTarget = mock.mockFunction('getTarget');
      const canSeeCount = mock.mockFunction('canSeeCount');
      getTarget.returned([]);
      canSeeCount.returned(0);
      getTarget.returned([]);
      canSeeCount.returned(10);
      exp(target.isNeedScrolelr()).to.be.false;
      exp(target.isNeedScrolelr()).to.be.false;
    });
    mount(<Target></Target>);
  });
  it('isNeedScrolelr is false 可见的数大于总数', () => {
    const Menus = () => {
      return <div>1</div>;
    };

    const Target = createTestComponent(ThrottleScroller(Menus, 20), target => {
      const mock = mockObject.create(target);
      const getTarget = mock.mockFunction('getTarget');
      const canSeeCount = mock.mockFunction('canSeeCount');
      getTarget.returned([]);
      canSeeCount.returned(5);
      exp(target.isNeedScrolelr()).to.be.false;
    });
    mount(<Target></Target>);
  });
  it('isNeedScrolelr is true 可见的数小于总数', () => {
    const Menus = () => {
      return <div>1</div>;
    };

    const Target = createTestComponent(ThrottleScroller(Menus, 20), target => {
      const mock = mockObject.create(target);
      const getTarget = mock.mockFunction('getTarget');
      const canSeeCount = mock.mockFunction('canSeeCount');
      getTarget.returned([1, 2, 3, 4, 5, 6,]);
      canSeeCount.returned(5);
      exp(target.isNeedScrolelr()).to.be.true;
    });
    mount(<Target></Target>);
  });

  it('fetchViewSize', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const height = 500;
    const theme = () => ({ height, });
    const Target = createTestComponent(ThrottleScroller(Menus, 20), target => {
      exp(target.fetchViewSize()).to.be.equal(height);
    });
    mount(<Target getTheme={theme}></Target>);
  });
  it('fetchViewSize default', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const height = 250;
    const Target = createTestComponent(ThrottleScroller(Menus, 20), target => {
      exp(target.fetchViewSize()).to.be.equal(height);
    });
    mount(<Target></Target>);
  });
  it('getStart', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const height = 500;
    const theme = () => ({ height, });
    const canSee = height / 20;
    const menuItemHeight = 20;
    const Target = createTestComponent(ThrottleScroller(Menus, menuItemHeight), target => {

      exp(target.getStart({}, { start: 100, }), 'is state.start').to.be.equal(100);
      exp(target.getStart({ start: 1000, }, { start: 100, }), 'is props.start').to.be.equal(1000 - canSee + 1);

      exp(target.getStart({}, {}), 'is empty 0').to.be.equal(0);

    });
    mount(<Target getTheme={theme} data={{ length: 1000, }}></Target>);
  });

  it('fetchTotalSize', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const height = 500;
    const theme = () => ({ height, });
    const itemHeigth = 20;
    const Target = createTestComponent(ThrottleScroller(Menus, itemHeigth), target => {
      const mock = mockObject.create(target);
      const getTarget = mock.mockFunction('getTarget');
      const items = [1, 2, 3, 4, 54, 5,];
      getTarget.returned(items);
      exp(target.fetchTotalSize()).to.be.equal(itemHeigth * items.length);
    });
    mount(<Target getTheme={theme}></Target>);
  });


  it('canSeeCount', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const itemHeight = 20;
    const Target = createTestComponent(ThrottleScroller(Menus, itemHeight), target => {
      const mock = mockObject.create(target);
      const fetchViewSize = mock.mockFunction('fetchViewSize');
      fetchViewSize.returned(100);
      exp(target.canSeeCount()).to.be.equal(5);

      fetchViewSize.returned(0);
      exp(target.canSeeCount()).to.be.equal(0);

      fetchViewSize.returned(-100);
      exp(target.canSeeCount()).to.be.equal(0);

      fetchViewSize.returned(81);
      exp(target.canSeeCount()).to.be.equal(5);

      fetchViewSize.returned(70);
      exp(target.canSeeCount()).to.be.equal(4);

      fetchViewSize.returned(69);
      exp(target.canSeeCount()).to.be.equal(4);

    });
    mount(<Target></Target>);
  });
  it('canSeeCount itemHeight =0', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const Target = createTestComponent(ThrottleScroller(Menus, 0), target => {
      const mock = mockObject.create(target);
      const fetchViewSize = mock.mockFunction('fetchViewSize');
      fetchViewSize.returned(100);
      exp(target.canSeeCount()).to.be.equal(0);

      fetchViewSize.returned(0);
      exp(target.canSeeCount()).to.be.equal(0);

      fetchViewSize.returned(-100);
      exp(target.canSeeCount()).to.be.equal(0);

      fetchViewSize.returned(81);
      exp(target.canSeeCount()).to.be.equal(0);

      fetchViewSize.returned(70);
      exp(target.canSeeCount()).to.be.equal(0);

      fetchViewSize.returned(69);
      exp(target.canSeeCount()).to.be.equal(0);

    });
    mount(<Target></Target>);
  });
  it('canSeeCount itemHeight <0', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const Target = createTestComponent(ThrottleScroller(Menus, -100), target => {
      const mock = mockObject.create(target);
      const fetchViewSize = mock.mockFunction('fetchViewSize');
      fetchViewSize.returned(100);
      exp(target.canSeeCount()).to.be.equal(0);

      fetchViewSize.returned(0);
      exp(target.canSeeCount()).to.be.equal(0);

      fetchViewSize.returned(-100);
      exp(target.canSeeCount()).to.be.equal(0);

      fetchViewSize.returned(81);
      exp(target.canSeeCount()).to.be.equal(0);

      fetchViewSize.returned(70);
      exp(target.canSeeCount()).to.be.equal(0);

      fetchViewSize.returned(69);
      exp(target.canSeeCount()).to.be.equal(0);

    });
    mount(<Target></Target>);
  });
  it('fetchStart&End', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const itemHeight = 20;
    const Target = createTestComponent(ThrottleScroller(Menus, itemHeight), target => {
      const mock = mockObject.create(target);
      const canSeeCount = mock.mockFunction('canSeeCount');
      const getTarget = mock.mockFunction('getTarget');
      getTarget.returned([]);
      exp(target.fetchEnd(3)).to.be.equal(0);

      getTarget.forever([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,]);
      canSeeCount.returned(5);
      exp(target.fetchEnd(3)).to.be.equal(8);

      canSeeCount.returned(5);
      exp(target.fetchEnd(4)).to.be.equal(9);

      canSeeCount.returned(5);
      exp(target.fetchEnd(5)).to.be.equal(10);

      canSeeCount.returned(5);
      exp(target.fetchEnd(100)).to.be.equal(11);

      canSeeCount.returned(5);
      exp(target.fetchEnd(-1)).to.be.equal(5);

      canSeeCount.returned(0);
      exp(target.fetchEnd(5)).to.be.equal(5);
    });
    mount(<Target></Target>);
  });


  it('没有滚动条', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const Target = createTestComponent(ThrottleScroller(Menus, 20), target => {
      const mock = mockObject.create(target);
      mock.mockFunction('isNeedScrolelr').returned(false);
    });
    const cmp = mount(<Target></Target>);
    cmp.render();
    exp(cmp.html()).to.be.equal('<div>1</div>');
  });

  it('有滚动条 滚动条的属性是否正确 scroler.type is default y', () => {
    createExistScrollerCase();
  });

  it('有滚动条 滚动条的属性是否正确 scroler.type is x', () => {
    createExistScrollerCase('x');

  });

  function createExistScrollerCase (type: ?string) {
    const Menus = () => {
      return <div>1</div>;
    };
    let _this: Object = {};
    const throttleScroller = ThrottleScroller(Menus, 20);
    const viewSize = 101,
      totalSize = 102;
    const Target = createTestComponent(throttleScroller, target => {
      _this = target;
      const mock = mockObject.create(target);
      mock.mockFunction('fetchViewSize').forever(viewSize);
      mock.mockFunction('fetchTotalSize').returned(totalSize);
      mock.mockFunction('isNeedScrolelr').mock(() => {
        return true;
      });
    });
    let cmp = mount(<Target type={type}></Target>);

    exp(_this.scroller).to.be.undefined;

    cmp = cmp.setState({ start: 1, });
    exp(cmp.find(SVScroller).length).to.be.equal(1);
    exp(cmp.find(SVScroller).props().onChange).to.be.equal(_this.onScroller);
    exp(cmp.find(SVScroller).props()).to.be.eql({
      viewSize,
      totalSize,
      value: 0,
      type: type ? type : 'y',
      onChange: _this.onScroller,
      throttle: 100,
      step: 1,
    });
    exp(_this.scroller).to.be.not.null;

    // onWheel
    const scroller = {};
    const mock = mockObject.create(_this);
    const mockScroller = mockObject.create(scroller);
    mock.mockVar('scroller').forever(scroller);

    const mockFunction = mockScroller.mockFunction('onWheel');
    mockFunction.returned(true);
    const event = { target: 'ligx', };
    _this.onWheel(event);
    exp(mockFunction.getCallContext(0)).to.be.equal(scroller);
    exp(mockFunction.getCallArgs(0)).to.be.eql([event,]);
  }

  it('Target props & start & end', () => {

    const Menus = () => {
      return <div>1</div>;
    };
    const start = 100;
    const end = 100;

    let _this: Object = {};
    const itemH = 20;
    const throttleScroller = ThrottleScroller(Menus, itemH);
    const viewSize = 101,
      totalSize = 102;
    const Target = createTestComponent(throttleScroller, target => {
      _this = target;
      const mock = mockObject.create(target);
      mock.mockFunction('fetchViewSize').forever(viewSize);
      mock.mockFunction('fetchTotalSize').forever(totalSize);
      mock.mockFunction('fetchEnd').returned(end);

      mock.mockFunction('isNeedScrolelr').mock(() => {
        return true;
      });
    });
    const canSee = Math.floor(viewSize / itemH);
    let cmp = mount(<Target data={{ length: 1000, }}></Target>);

    cmp = cmp.setState({ start, });
    let menuProps = cmp.find(Menus).props();
    exp(menuProps.end).to.be.equal(end);
    exp(menuProps.start).to.be.equal(0);
    // 测试滚动
    _this.onScroller(100);
    cmp.setState({ start, });
    menuProps = cmp.find(Menus).props();
    exp(menuProps.start).to.be.equal(5);
    exp(menuProps.end).to.be.equal(5 + canSee + 1);
  });

  it('需要滚动条 HTML结构', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const itemH = 20;
    const Target = ThrottleScroller(Menus, itemH);
    const theme = () => ({ height: 50, width: 300, });
    const data = [1, 2, 3, 4, 5, 6, 7, 8, 9,];
    expect(renderer.create(<Target getTheme={theme}
                                   data={data}></Target>).toJSON()).toMatchSnapshot();
  });

  it('需要滚动条 HTML结构 theme is default data is 10 len', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const itemH = 20;
    const Target = ThrottleScroller(Menus, itemH);
    const data = [1, 2, 3, 4, 5, 6, 7, 8, 9,];
    expect(renderer.create(<Target data={data}></Target>).toJSON()).toMatchSnapshot();
  });
  it('需要滚动条 HTML结构 theme is default data is 100 len', () => {
    const Menus = () => {
      return <div>1</div>;
    };
    const itemH = 20;
    const Target = ThrottleScroller(Menus, itemH);
    const data = [];
    for (let i = 0; i < 100; i++) {
      data.push(i);
    }
    expect(renderer.create(<Target data={data}></Target>).toJSON()).toMatchSnapshot();
  });
  it('需要滚动条 HTML结构 children', () => {
    const Item = (props: Object) => {
      return <div>{props.title}</div>;
    };
    const items = [];
    for (let i = 0; i < 100; i++) {
      items.push(<Item key={`key${i}`} title={`title${i}`}/>);
    }
    const Menus = props => {
      return <div>{props.children}</div>;
    };

    const itemH = 20;
    const Target = ThrottleScroller(Menus, itemH);

    expect(renderer.create(<Target>{items}</Target>).toJSON()).toMatchSnapshot();
  });
});
