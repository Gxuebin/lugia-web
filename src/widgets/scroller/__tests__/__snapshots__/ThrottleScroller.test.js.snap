// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThrottleScroller 需要滚动条 HTML结构 1`] = `
.c4 {
  position: relative;
  background: #fbfbfb;
  width: 1.6666666666666667em;
  height: 25em;
  z-index: 996;
  width: 1em;
}

.c5 {
  position: absolute;
  top: 0;
  left: 0;
  cursor: pointer;
  text-align: center;
  border-radius: 0.6666666666666667em;
  background: #c2c2c2;
  color: #fff;
  font-size: 1.2rem;
  line-height: 2.5em;
  -webkit-transition: all 0.4s;
  transition: all 0.4s;
  width: 0.6666666666666667em;
  margin-left: 0.16666666666666669em;
  margin-right: 0.16666666666666669em;
}

.c5:hover {
  background-color: #7a7a7a;
}

.c2 {
  width: 24em;
  position: absolute;
  display: inline-block;
}

.c1 {
  overflow: hidden;
  height: 4.166666666666667em;
  width: 25em;
  position: relative;
}

.c0:hover .c6 {
  opacity: 1;
}

.c3 {
  width: 24em;
  position: absolute;
  display: inline-block;
  left: 24em;
  width: 1em;
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c0:hover .c3 {
  opacity: 1;
}

<div
  className="c0 c1"
  onWheel={[Function]}
>
  <div
    className="c2"
  >
    <div>
      1
    </div>
  </div>
  <div
    className="c3"
  >
    <div
      className="c4"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      onWheel={[Function]}
      style={
        Object {
          "height": "4.166666666666667em",
        }
      }
    >
      <div
        className="c5"
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "height": "1.1666666666666667em",
            "top": "0em",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`ThrottleScroller 需要滚动条 HTML结构 children 1`] = `
.c4 {
  position: relative;
  background: #fbfbfb;
  width: 1.6666666666666667em;
  height: 25em;
  z-index: 996;
  width: 1em;
}

.c5 {
  position: absolute;
  top: 0;
  left: 0;
  cursor: pointer;
  text-align: center;
  border-radius: 0.6666666666666667em;
  background: #c2c2c2;
  color: #fff;
  font-size: 1.2rem;
  line-height: 2.5em;
  -webkit-transition: all 0.4s;
  transition: all 0.4s;
  width: 0.6666666666666667em;
  margin-left: 0.16666666666666669em;
  margin-right: 0.16666666666666669em;
}

.c5:hover {
  background-color: #7a7a7a;
}

.c2 {
  width: 19.833333333333336em;
  position: absolute;
  display: inline-block;
}

.c1 {
  overflow: hidden;
  height: 20.833333333333336em;
  width: 20.833333333333336em;
  position: relative;
}

.c3 {
  width: 19.833333333333336em;
  position: absolute;
  display: inline-block;
  left: 19.833333333333336em;
  width: 1em;
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c0:hover .c3 {
  opacity: 1;
}

.c0:hover .c6 {
  opacity: 1;
}

<div
  className="c0 c1"
  onWheel={[Function]}
>
  <div
    className="c2"
  >
    <div>
      <div>
        title0
      </div>
      <div>
        title1
      </div>
      <div>
        title2
      </div>
      <div>
        title3
      </div>
      <div>
        title4
      </div>
      <div>
        title5
      </div>
      <div>
        title6
      </div>
      <div>
        title7
      </div>
      <div>
        title8
      </div>
      <div>
        title9
      </div>
      <div>
        title10
      </div>
      <div>
        title11
      </div>
      <div>
        title12
      </div>
      <div>
        title13
      </div>
      <div>
        title14
      </div>
      <div>
        title15
      </div>
      <div>
        title16
      </div>
      <div>
        title17
      </div>
      <div>
        title18
      </div>
      <div>
        title19
      </div>
      <div>
        title20
      </div>
      <div>
        title21
      </div>
      <div>
        title22
      </div>
      <div>
        title23
      </div>
      <div>
        title24
      </div>
      <div>
        title25
      </div>
      <div>
        title26
      </div>
      <div>
        title27
      </div>
      <div>
        title28
      </div>
      <div>
        title29
      </div>
      <div>
        title30
      </div>
      <div>
        title31
      </div>
      <div>
        title32
      </div>
      <div>
        title33
      </div>
      <div>
        title34
      </div>
      <div>
        title35
      </div>
      <div>
        title36
      </div>
      <div>
        title37
      </div>
      <div>
        title38
      </div>
      <div>
        title39
      </div>
      <div>
        title40
      </div>
      <div>
        title41
      </div>
      <div>
        title42
      </div>
      <div>
        title43
      </div>
      <div>
        title44
      </div>
      <div>
        title45
      </div>
      <div>
        title46
      </div>
      <div>
        title47
      </div>
      <div>
        title48
      </div>
      <div>
        title49
      </div>
      <div>
        title50
      </div>
      <div>
        title51
      </div>
      <div>
        title52
      </div>
      <div>
        title53
      </div>
      <div>
        title54
      </div>
      <div>
        title55
      </div>
      <div>
        title56
      </div>
      <div>
        title57
      </div>
      <div>
        title58
      </div>
      <div>
        title59
      </div>
      <div>
        title60
      </div>
      <div>
        title61
      </div>
      <div>
        title62
      </div>
      <div>
        title63
      </div>
      <div>
        title64
      </div>
      <div>
        title65
      </div>
      <div>
        title66
      </div>
      <div>
        title67
      </div>
      <div>
        title68
      </div>
      <div>
        title69
      </div>
      <div>
        title70
      </div>
      <div>
        title71
      </div>
      <div>
        title72
      </div>
      <div>
        title73
      </div>
      <div>
        title74
      </div>
      <div>
        title75
      </div>
      <div>
        title76
      </div>
      <div>
        title77
      </div>
      <div>
        title78
      </div>
      <div>
        title79
      </div>
      <div>
        title80
      </div>
      <div>
        title81
      </div>
      <div>
        title82
      </div>
      <div>
        title83
      </div>
      <div>
        title84
      </div>
      <div>
        title85
      </div>
      <div>
        title86
      </div>
      <div>
        title87
      </div>
      <div>
        title88
      </div>
      <div>
        title89
      </div>
      <div>
        title90
      </div>
      <div>
        title91
      </div>
      <div>
        title92
      </div>
      <div>
        title93
      </div>
      <div>
        title94
      </div>
      <div>
        title95
      </div>
      <div>
        title96
      </div>
      <div>
        title97
      </div>
      <div>
        title98
      </div>
      <div>
        title99
      </div>
    </div>
  </div>
  <div
    className="c3"
  >
    <div
      className="c4"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      onWheel={[Function]}
      style={
        Object {
          "height": "20.833333333333336em",
        }
      }
    >
      <div
        className="c5"
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "height": "2.5833333333333335em",
            "top": "0em",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`ThrottleScroller 需要滚动条 HTML结构 theme is default data is 10 len 1`] = `
<div>
  1
</div>
`;

exports[`ThrottleScroller 需要滚动条 HTML结构 theme is default data is 100 len 1`] = `
.c4 {
  position: relative;
  background: #fbfbfb;
  width: 1.6666666666666667em;
  height: 25em;
  z-index: 996;
  width: 1em;
}

.c5 {
  position: absolute;
  top: 0;
  left: 0;
  cursor: pointer;
  text-align: center;
  border-radius: 0.6666666666666667em;
  background: #c2c2c2;
  color: #fff;
  font-size: 1.2rem;
  line-height: 2.5em;
  -webkit-transition: all 0.4s;
  transition: all 0.4s;
  width: 0.6666666666666667em;
  margin-left: 0.16666666666666669em;
  margin-right: 0.16666666666666669em;
}

.c5:hover {
  background-color: #7a7a7a;
}

.c2 {
  width: 19.833333333333336em;
  position: absolute;
  display: inline-block;
}

.c1 {
  overflow: hidden;
  height: 20.833333333333336em;
  width: 20.833333333333336em;
  position: relative;
}

.c3 {
  width: 19.833333333333336em;
  position: absolute;
  display: inline-block;
  left: 19.833333333333336em;
  width: 1em;
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c0:hover .c3 {
  opacity: 1;
}

.c0:hover .c6 {
  opacity: 1;
}

<div
  className="c0 c1"
  onWheel={[Function]}
>
  <div
    className="c2"
  >
    <div>
      1
    </div>
  </div>
  <div
    className="c3"
  >
    <div
      className="c4"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      onWheel={[Function]}
      style={
        Object {
          "height": "20.833333333333336em",
        }
      }
    >
      <div
        className="c5"
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "height": "2.5833333333333335em",
            "top": "0em",
          }
        }
      />
    </div>
  </div>
</div>
`;
